src/header/dib_header.rs

```diff
@@ -0,0 +1,142 @@
+        // The header size in the BMP includes its own u32, so we strip it out by subtracting 4
+        // bytes to get the right final offset to the end of the header.
+        let (input, dib_header_data) = length_data(map(le_u32, |len| len - 4))(input)?;
+
+        // Add 4 back on so the constants remain the correct size relative to the BMP
+        // documentation/specs.
+        let header_type = match dib_header_data.len() + 4 {
+            DIB_V3_HEADER_SIZE => HeaderType::V3,
+            DIB_V4_HEADER_SIZE => HeaderType::V4,
+            DIB_V5_HEADER_SIZE => HeaderType::V5,
+            DIB_INFO_HEADER_SIZE => HeaderType::Info,
+            _ => {
+                return Err(nom::Err::Failure(nom::error::Error::from_error_kind(
+                    dib_header_data,
+                    ErrorKind::LengthValue,
+                )))
+            }
+        };
+
+        // Fields common to all DIB variants
+        let (dib_header_data, image_width) = le_u32(dib_header_data)?;
+        let (dib_header_data, image_height) = le_i32(dib_header_data)?;
+        let (dib_header_data, _color_planes) = le_u16(dib_header_data)?;
+        let (dib_header_data, bpp) = Bpp::parse(dib_header_data)?;
+
+        // Extra fields defined by DIB variants
+        // Variants are described in
+        // <https://www.liquisearch.com/bmp_file_format/file_structure/dib_header_bitmap_information_header>
+        // and <https://docs.microsoft.com/en-us/windows/win32/gdi/bitmap-header-types>
+        let (dib_header_data, compression_method) = CompressionMethod::parse(dib_header_data)?;
+        let (dib_header_data, image_data_len) = le_u32(dib_header_data)?;
+        let (dib_header_data, _pels_per_meter_x) = le_u32(dib_header_data)?;
+        let (dib_header_data, _pels_per_meter_y) = le_u32(dib_header_data)?;
+        let (dib_header_data, colors_used) = le_u32(dib_header_data)?;
+        let (dib_header_data, _colors_important) = le_u32(dib_header_data)?;
+
+        let (_dib_header_data, channel_masks) = match header_type {
+            HeaderType::V3 | HeaderType::V4 | HeaderType::V5
+                if compression_method == CompressionMethod::Bitfields =>
+            {
+                let (dib_header_data, mask_red) = le_u32(dib_header_data)?;
+                let (dib_header_data, mask_green) = le_u32(dib_header_data)?;
+                let (dib_header_data, mask_blue) = le_u32(dib_header_data)?;
+                let (dib_header_data, mask_alpha) = le_u32(dib_header_data)?;
+
+                (
+                    dib_header_data,
+                    Some(ChannelMasks {
+                        red: mask_red,
+                        green: mask_green,
+                        blue: mask_blue,
+                        alpha: mask_alpha,
+                    }),
+                )
+            }
+            HeaderType::Info => (dib_header_data, None),
+            _ => (dib_header_data, None),
+        };
```

Comment:
Maybe this is a pointless refactoring, but IMO its more readable. Requires the changes to `HeaderType`.

```suggestion
        let (_dib_header_data, channel_masks) = if header_type.is_at_least(HeaderType::V3)
            && compression_method == CompressionMethod::Bitfields
        {
            let (dib_header_data, mask_red) = le_u32(dib_header_data)?;
            let (dib_header_data, mask_green) = le_u32(dib_header_data)?;
            let (dib_header_data, mask_blue) = le_u32(dib_header_data)?;
            let (dib_header_data, mask_alpha) = le_u32(dib_header_data)?;

            (
                dib_header_data,
                Some(ChannelMasks {
                    red: mask_red,
                    green: mask_green,
                    blue: mask_blue,
                    alpha: mask_alpha,
                }),
            )
        } else {
            (dib_header_data, None)
        };
```